### PR TITLE
Framework: Allow masterbar to be hidden for mobile app webviews

### PR DESCRIPTION
**Pull request by seear**
Add a modifier CSS class `has-no-masterbar` that sets masterbar opacity to 0. Control the presence of this class with `masterbarVisibility` app ui state.

Part of jetpack connection from the mobile app, where we want to hide as many links as possible to keep customer in the connection flow. For further details see p7rd6c-1aP-p2 and #20955.

masterbar_mobile

**Why do we need to be able to hide the masterbar?**
So that the mobile apps can use parts of Calypso in a webview without the user being able to click around to other parts of Calypso. Specifically, this will be used for connecting Jetpack sites. #21408 makes use of the new state added here to hide the masterbar for the mobile app connection flow.

**Why can't we do this locally in the jetpack connect code?**
The Layout components that render the master are at the top level of Calypso, so we need to add app-wide state to control the presence of a modifier class.

**What risks does this change introduce?**
Although we can fully test the jetpack connection flow, this change allows the masterbar to be hidden anywhere in Calypso. There could be layout or other issues if any future code takes advantage of this.

**Is the masterbar hidden currently in any other parts of Calypso?**
The editor hides the masterbar by setting opacity to 0 based on the presence of `.is-section-post-editor`, see #20326.

## Testing
* Hide/show the masterbar from the console with:
```
dispatch( { type: 'MASTERBAR_TOGGLE_VISIBILITY', isVisible: false } );
dispatch( { type: 'MASTERBAR_TOGGLE_VISIBILITY', isVisible: true } );
```

### CODE REVIEW

File: client/state/ui/selectors.js [hunk 1]
```diff
@@ -134,3 +134,7 @@ export function hasSidebar( state ) {
 	}
 	return get( getSection( state ), 'secondary', true );
 }
+
+export function shouldHideMasterbar( state ) {
+	return get( state.ui, 'hideMasterbar', false );
+}
```
Comment (sirreal) [inline client/state/ui/selectors.js]: I don't have strong feelings (actually I'm a bit conflicted), but note that [there's a guideline](https://github.com/Automattic/wp-calypso/blob/master/client/state/selectors/README.md#state-selectors):

> [`client/state/selectors`] contains all available state selectors. Each file includes a single default exported function which can be used as a helper in retrieving derived data from the global state tree.

File: client/layout/masterbar/style.scss [hunk 2]
```diff
@@ -489,3 +489,8 @@ $autobar-height: 20px;
 		transform: translateY( -48px );
 	}
 }
+
+.hide-masterbar .masterbar {
+	opacity: 0;
+	pointer-events: none;
```
Comment (sirreal) [inline client/layout/masterbar/style.scss]: Thinking out loud-

`visibility: hidden` should cover both of these rules. It does not have a smooth transition which may or may not have a use case. Probably good to leave this as is.

File: client/state/ui/selectors.js (hunk 1 above)
Comment (seear) [reply client/state/ui/selectors.js]: I'm always confused when there is an existing file containing all the selectors. Happy to move this new one to its own file, but I always have in mind that I would be confused to find selectors for an area split across one big file and a couple of individual ones.

**Comment by lsinger** [comment]
In #20326 we want to smoothly transition to the hidden Masterbar and replace it with the Editor's toolbar / "ground control." In this case I don't think any transition is planned, or is it? So why not just use `display: none` and move things up? As it is, there's now unused space on top, with a slightly darker color as well.

**Comment by seear** [comment]
> So why not just use display: none and move things up? As it is, there's now unused space on top, with a slightly darker color as well.

Good point. Correct that we don't need a transition. I'll play with that, though I think leaving the space at the top will give better balance. Take a look at the shots in #21408 for the use case.

The darker color in the gif in the description is just a gif artifact I think.

**Comment by lsinger** [comment]
> The darker color in the gif in the description is just a gif artifact I think.

Confirmed. 

It does look odd when opening the sidebar, though:

no-masterbar

Is this something that could be encountered in the scenarios this would be used in?

**Comment by seear** [comment]
> It does look odd when opening the sidebar... Is this something that could be encountered in the scenarios this would be used in?

No, the jetpack connect flow has no sidebar, and we are only interested in mobile widths.

How about we make this explicit by using `.has-no-sidebar.hide-masterbar`?

**Comment by sirreal** [comment]
> How about we make this explicit by using .has-no-sidebar.hide-masterbar?

I'm not sure if there's much use for this beyond our particular case. I'm inclined to say <abbr title="You ain't gonna need it">YAGNI</abbr> on this one. Let's not over engineer and if others folks have need, they can figure out adequate solutions.

Principle of least surprise would also tell me that hiding the masterbar _does not_ main hide the sidebar as well.

**Comment by seear** [comment]
> Principle of least surprise would also tell me that hiding the masterbar does not mean hide the sidebar as well.

Oh, sorry, I meant only hide the masterbar when the (existing) `.has-no-sidebar` class is present, by changing the css to:
```
.has-no-sidebar.hide-masterbar .masterbar {
	opacity: 0;
	pointer-events: none;
}
```

I don't disagree with your point though, it is adding complication.

**Comment by seear** [comment]
Applied changes from review of #21408 and updated testing instructions to:
```
dispatch( { type: 'MASTERBAR_TOGGLE_VISIBILITY', isVisible: false } );
dispatch( { type: 'MASTERBAR_TOGGLE_VISIBILITY', isVisible: true } );
```

**Comment by seear** [comment]
Thanks for the reviews all, really appreciate it.

**Pull request merged**